### ISSUE TITLE
Speed up TCO when there are default param values

### ISSUE THREAD

**Issue by jfairbank** [babel/babel#1789]
While fiddling with TCO in Babel, I decided to compare a TCO factorial function with a normal recursive factorial function. I noticed that the runtime was significantly worse for the TCO. I thought this was odd because I know TCO has been well touted in Babel. Then, I noticed how I was writing my TCO function and wrote a couple different versions. Surprisingly, my alternate versions ran significantly better; one of them even ran better than the normal recursive implementation.

What I discovered is that TCO functions with default param values appear slower. Eyeballing the codebase (which I'm not terribly familiar with), I believe it's because default params cause the `arguments` variable to be used. In TCO, this leads to recreating an array for every "recursive" call.

Here's the original code:

```js
// TCO with default accum param
function factorialTco1(n, accum = 1) {
  if (n < 2) return accum;
  return factorialTco1(n - 1, n * accum);
}
```

And transpiled code:

```js
function factorialTco1(_x2) {
  var _arguments = arguments;
  var _again = true;

  _function: while (_again) {
    var n = _x2;
    accum = undefined;
    _again = false;
    var accum = _arguments[1] === undefined ? 1 : _arguments[1];

    if (n < 2) return accum;
    _arguments = [_x2 = n - 1, n * accum];
    _again = true;
    continue _function;
  }
}
```

Checking the param manually and setting the default value seems to yield significantly better performance because we don't mess with `arguments` and recreating the array.

```js
// TCO that checks if the accum has been set and
// give a default value if not
function factorialTco3(n, accum) {
  if (accum === undefined) accum = 1;
  if (n < 2) return accum;
  return factorialTco3(n - 1, n * accum);
}
```

Transpiled:

```js
function factorialTco3(_x5, _x6) {
  var _again3 = true;

  _function3: while (_again3) {
    var n = _x5,
        accum = _x6;
    _again3 = false;

    if (accum === undefined) accum = 1;
    if (n < 2) return accum;
    _x5 = n - 1;
    _x6 = n * accum;
    _again3 = true;
    continue _function3;
  }
}
```

Assuming I set it up right, the corresponding jsPerf is here: http://jsperf.com/babel-tco.

---

All that to ask: is it feasible and worthwhile to change the implementation for default param values? For small inputs and most use cases this may not matter because the performance difference between the TCO and regular recursive implementations isn't very large. I figured I would still bring it to your attention if you do find it worthwhile to optimize.

Obviously if performance is important to me, then I can just write it the alternate way, but I do appreciate the slightly more expressive and succinct way with default param values.

I'd be more than happy to PR this up as well, but I confess I would need some direction on where to get started with the codebase.

**Comment by sebmck** [comment]
That changes the arity of the function though and isn't spec compliant.

**Comment by sebmck** [comment]
To elaborate:

```javascript
function foo(bar = 1) {

}

foo.length;
```

should be `0` and not `1`.

**Comment by jfairbank** [comment]
I honestly never thought about the arity and how that is affected by default params (which are optional parameters), so I'm having a duh moment now. I admit I'm ignorant on the spec as well, so I figured there might be compelling reasons why this implementation is what it is. I just glanced at part of the spec for function declarations, so this makes sense.

I'm still treading in some ignorance on my part, but is there any reason why TCO can't be optimized nonetheless? I guess I'm asking what would break in TCO if we held variable references to optional params instead of stuffing their values back in an array?

Again this really isn't a *huge* deal because most use cases won't necessitate this optimization. I completely understand if you don't find this worthwhile or even doable given the spec.

Thanks for entertaining my ignorance either way :)

**Comment by sebmck** [comment]
>  I guess I'm asking what would break in TCO if we held variable references to optional params instead of stuffing their values back in an array?

What would that look like? What about functions that references `arguments` to `apply` a function or something that are indistinguishable?

**Comment by jfairbank** [comment]
Yes, checking for explicit `arguments` introduces a whole new problem too if we used variable references instead.

I was going to suggest mutating `_arguments` as a possible solution too -- I'm not sure yet what that could break -- but it appears that that doesn't provide any better performance from what I just tested. That's probably due to the use of `arguments` in the function.

At this point, I may be going off on a rabbit trail, so this may not be worthwhile. Thank you for working with me through this @sebmck.
